File: pollux/lib/core/src/main/scala/io/iohk/atala/pollux/core/model/VerificationPolicy.scala
package io.iohk.atala.pollux.core.model

import zio.{Clock, Random, ZIO}

import java.time.{OffsetDateTime, ZoneId, ZoneOffset}
import java.util.UUID

case class VerificationPolicy(
    id: UUID,
    name: String,
    description: String,
    createdAt: OffsetDateTime,
    updatedAt: OffsetDateTime,
    constrains: Seq[VerificationPolicyConstraint],
    nonce: Int
)

object VerificationPolicy {
  def make(name: String, description: String, constraints: Seq[VerificationPolicyConstraint], nonce: Int = 0) =
    for {
      id <- Random.nextUUID
      ts <- Clock.currentDateTime.map(_.atZoneSameInstant(ZoneOffset.UTC).toOffsetDateTime)
    } yield VerificationPolicy(
      id,
      name,
      description,
      createdAt = ts,
      updatedAt = ts,
      constrains = constraints,
      nonce = nonce
    )
}
sealed trait VerificationPolicyConstraint

case class CredentialSchemaAndTrustedIssuersConstraint(
    schemaId: String,
    trustedIssuers: Seq[String]
) extends VerificationPolicyConstraint

File: tests/e2e-tests/src/main/kotlin/api_models/VerificationPolicy.kt
package api_models

data class VerificationPolicy(
    var id: String? = null,
    var nonce: String? = null,
    var name: String? = null,
    var description: String? = null,
    var constraints: List<Constraint>? = null,
    var createdAt: String? = null,
    var updatedAt: String? = null,
    var kind: String? = null,
    var self: String? = null,
)

data class VerificationPolicyInput(
    var name: String? = null,
    var description: String? = null,
    var constraints: List<Constraint>? = null,
)

data class Constraint(
    var schemaId: String? = null,
    var trustedIssuers: List<String>? = null,
)
